### PR TITLE
add: ensure envoys can only connect after caches have been synced

### PR DESCRIPTION
**Pull request by ramaraochavali**
This PR does

- Adds a readiness probe for Istiod that ensures Istiod is ready only after caches are synced.
- Rejects connections before cache sync. See https://github.com/istio/istio/issues/25495#issuecomment-661846453

There is still an issue with sync that this PR https://github.com/istio/istio/pull/24339 will address.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn @hzxuzhonghu PTAL

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -305,6 +305,10 @@ func NewServer(args *PilotArgs) (*Server, error) {
 		})
 	}
 
+	s.addReadinessProbe("discovery", func() (bool, error) {
+		return s.cachesSynced(), nil
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:309]: can check `IsServerReady` here

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -243,6 +243,18 @@ func (s *DiscoveryServer) processRequest(discReq *discovery.DiscoveryRequest, co
...
 func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	// Check if server is ready to accept clients and process new requests.
+	// Currently ready means caches have been synced and hence can build
+	// clusters correctly. Without this check, InitContext() call below would
+	// initialize with empty config, leading to reconnected Envoys loosing
+	// configuration. This is an additional safety check inaddition to adding
+	// cachesSynced logic to readiness probe to handle cases where kube-proxy
+	// ip tables update latencies.
+	// See https://github.com/istio/istio/issues/25495.
+	if !s.IsServerReady() {
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:254]: looks good

**Pull request merged**